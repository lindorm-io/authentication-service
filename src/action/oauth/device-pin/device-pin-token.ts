import Joi from "@hapi/joi";
import { IKoaAuthContext, ICreateTokensData } from "../../../typing";
import { InvalidDeviceError } from "../../../error";
import { JOI_EMAIL, JOI_GRANT_TYPE } from "../../../constant";
import { authenticateSession, createTokens, findValidSession } from "../../../support";
import { stringComparison } from "@lindorm-io/core";
import { requestVerifyDevicePIN } from "../../../axios";

export interface IPerformDevicePINTokenOptions {
  codeVerifier: string;
  deviceId: string;
  deviceVerifier: string;
  grantType: string;
  pin: string;
  subject: string;
}

const schema = Joi.object({
  codeVerifier: Joi.string().required(),
  deviceId: Joi.string().guid().required(),
  deviceVerifier: Joi.string().required(),
  grantType: JOI_GRANT_TYPE,
  pin: Joi.string().required(),
  subject: JOI_EMAIL,
});

export const performDevicePINToken = (ctx: IKoaAuthContext) => {
  return async (options: IPerformDevicePINTokenOptions): Promise<ICreateTokensData> => {
    await schema.validateAsync(options);

    const { client, metadata, repository } = ctx;
    const { codeVerifier, deviceId, deviceVerifier, grantType, pin, subject } = options;
    const authMethodsReference = "pin";

    if (!stringComparison(deviceId, metadata.deviceId)) {
      throw new InvalidDeviceError(deviceId);
    }

    const session = await findValidSession(ctx)({
      codeVerifier,
      grantType,
      subject,
    });

    const account = await repository.account.find({ email: session.authorization.email });

    await requestVerifyDevicePIN({
      account,
      deviceVerifier,
      pin,
      session,
    });

    const authenticated = await authenticateSession(ctx)({
      account,
      session,
    });

    return createTokens(ctx)({
      account,
      authMethodsReference,
      client,
      responseType: session.authorization.responseType,
      session: authenticated,
    });
  };
};
