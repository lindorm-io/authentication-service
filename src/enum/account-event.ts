export enum AccountEvent {
  CREATED = "account_created",
  EMAIL_CHANGED = "account_email_changed",
  OTP_CHANGED = "account_otp_changed",
  PASSWORD_CHANGED = "account_password_changed",
  PERMISSION_CHANGED = "account_permission_changed",
}
