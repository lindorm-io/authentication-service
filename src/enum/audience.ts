export enum Audience {
  ACCESS = "access",
  AUTHORIZATION = "authorization",
  IDENTITY = "identity",
  MULTI_FACTOR = "multi-factor",
  REFRESH = "refresh",
}
