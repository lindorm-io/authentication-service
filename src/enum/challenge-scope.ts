export enum ChallengeScope {
  CHANGE = "change",
  SIGN_IN = "sign_in",
  RECOVERY = "recovery",
}
