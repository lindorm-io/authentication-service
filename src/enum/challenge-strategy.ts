export enum ChallengeStrategy {
  IMPLICIT = "implicit",
  PIN = "pin",
  SECRET = "secret",
}
