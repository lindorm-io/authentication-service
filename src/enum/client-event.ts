export enum ClientEvent {
  CREATED = "client_created",
  APPROVED_CHANGED = "client_approved_changed",
  DESCRIPTION_CHANGED = "client_description_changed",
  EMAIL_AUTHORIZATION_URI_CHANGED = "client_email_authorization_uri_changed",
  NAME_CHANGED = "client_name_changed",
  SECRET_CHANGED = "client_secret_changed",
}
