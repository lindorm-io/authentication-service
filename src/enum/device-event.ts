export enum DeviceEvent {
  CREATED = "device_created",
  PIN_CHANGED = "device_pin_changed",
  SECRET_CHANGED = "device_secret_changed",
  NAME_CHANGED = "device_name_changed",
}
