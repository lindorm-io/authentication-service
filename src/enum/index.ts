export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
