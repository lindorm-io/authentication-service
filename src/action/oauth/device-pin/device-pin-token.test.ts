import { GrantType, ResponseType } from "../../../enum";
import { getTestRepository, getTestAccount, resetStore } from "../../../test";
import { performDevicePINToken } from "./device-pin-token";

jest.mock("../../../support", () => ({
  authenticateSession: jest.fn(() => () => "session"),
  createTokens: jest.fn(() => () => "tokens"),
  findValidSession: jest.fn(() => () => ({
    authorization: {
      email: "email@lindorm.io",
      responseType: ResponseType.REFRESH,
    },
  })),
}));
jest.mock("../../../axios", () => ({
  verifyDevicePIN: jest.fn(),
}));

describe("performDevicePINToken", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      client: "client",
      metadata: {
        deviceId: "deb1abb0-a747-451e-bd82-7a2d89f950ac",
      },
      repository: await getTestRepository(),
    };

    await ctx.repository.account.create(getTestAccount("email@lindorm.io"));
  });

  afterEach(resetStore);

  test("should return tokens", async () => {
    await expect(
      performDevicePINToken(ctx)({
        codeVerifier: "codeVerifier",
        deviceId: "deb1abb0-a747-451e-bd82-7a2d89f950ac",
        deviceVerifier: "deviceVerifier",
        grantType: GrantType.REFRESH_TOKEN,
        pin: "pin",
        subject: "email@lindorm.io",
      }),
    ).resolves.toMatchSnapshot();
  });
});
