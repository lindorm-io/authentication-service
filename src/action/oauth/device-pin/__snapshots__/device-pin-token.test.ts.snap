// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performDevicePINToken should return tokens 1`] = `"tokens"`;
